compiler/src/dotty/tools/dotc/core/Definitions.scala

```diff
@@ -1353,38 +1454,56 @@ class Definitions {
   def isValueSubClass(sym1: Symbol, sym2: Symbol): Boolean =
     valueTypeEnc(sym2.asClass.name) % valueTypeEnc(sym1.asClass.name) == 0
 
-  lazy val specialErasure: SimpleIdentityMap[Symbol, ClassSymbol] =
-    SimpleIdentityMap.Empty[Symbol]
-      .updated(AnyClass, ObjectClass)
-      .updated(AnyValClass, ObjectClass)
-      .updated(SingletonClass, ObjectClass)
-      .updated(TupleClass, ObjectClass)
-      .updated(NonEmptyTupleClass, ProductClass)
+  lazy val specialErasure: SimpleIdentityMap[Symbol, ClassSymbol] = {
+    val idMap =
+      SimpleIdentityMap.Empty[Symbol]
+        .updated(AnyClass, ObjectClass)
+        .updated(AnyValClass, ObjectClass)
+        .updated(SingletonClass, ObjectClass)
+        .updated(TupleClass, ObjectClass)
+        .updated(NonEmptyTupleClass, ProductClass)
+    if (ctx.settings.YexplicitNulls.value) {
+      idMap.updated(RefEqClass, ObjectClass)
+    } else {
+      idMap
+    }
+  }
 
   // ----- Initialization ---------------------------------------------------
 
   /** Lists core classes that don't have underlying bytecode, but are synthesized on-the-fly in every reflection universe */
-  lazy val syntheticScalaClasses: List[TypeSymbol] = List(
-    AnyClass,
-    AnyRefAlias,
-    AnyKindClass,
-    andType,
-    orType,
-    RepeatedParamClass,
-    ByNameParamClass2x,
-    AnyValClass,
-    NullClass,
-    NothingClass,
-    SingletonClass,
-    EqualsPatternClass)
-
-  lazy val syntheticCoreClasses: List[Symbol] = syntheticScalaClasses ++ List(
+  def syntheticScalaClasses(implicit ctx: Context): List[TypeSymbol] = {
+    val synth = List(
+      AnyClass,
+      AnyRefAlias,
+      AnyKindClass,
+      andType,
+      orType,
+      RepeatedParamClass,
+      ByNameParamClass2x,
+      AnyValClass,
+      NullClass,
+      NothingClass,
+      SingletonClass,
+      EqualsPatternClass)
+
+    if (ctx.settings.YexplicitNulls.value) synth ++ List(JavaNullAlias, RefEqClass)
+    else synth
+  }
```

Comment:
For both `syntheticScalaClasses` and `syntheticCoreClasses` : What is the difference from using the `Definition.this.ctx`, so that we can keep `lazy val`?

Reply:
For context: `syntheticCoreClasses` is used on a hot patch in `TreeChecker` so it needs to be memoized, either in `Definitions` or in `TreeChecker`.

Reply:
I had missed this. Reverted. Thanks.